cumulus/client/bootnodes/src/advertisement.rs

```diff
@@ -0,0 +1,592 @@
+	/// Parachain side public addresses.
+	pub public_addresses: Vec<Multiaddr>,
+}
+
+/// Parachain bootnode advertisement service.
+pub struct BootnodeAdvertisement {
+	para_id: ParaId,
+	para_id_scale_compact: Vec<u8>,
+	relay_chain_interface: Arc<dyn RelayChainInterface>,
+	relay_chain_network: Arc<dyn NetworkService>,
+	current_epoch_key: Option<KademliaKey>,
+	next_epoch_key: Option<KademliaKey>,
+	current_epoch_publish_retry: Pin<Box<Fuse<Sleep>>>,
+	next_epoch_publish_retry: Pin<Box<Fuse<Sleep>>>,
+	request_receiver: async_channel::Receiver<IncomingRequest>,
+	parachain_network: Arc<dyn NetworkService>,
+	advertise_non_global_ips: bool,
+	parachain_genesis_hash: Vec<u8>,
+	parachain_fork_id: Option<String>,
+	public_addresses: Vec<Multiaddr>,
+}
+
+impl BootnodeAdvertisement {
+	/// Create a new bootnode advertisement service.
+	pub fn new(
+		BootnodeAdvertisementParams {
+			para_id,
+			relay_chain_interface,
+			relay_chain_network,
+			request_receiver,
+			parachain_network,
+			advertise_non_global_ips,
+			parachain_genesis_hash,
+			parachain_fork_id,
+			public_addresses,
+		}: BootnodeAdvertisementParams,
+	) -> Self {
+		// Discard `/p2p/<peer_id>` from public addresses on initialization to not generate warnings
+		// on every request for what is an operator mistake.
+		let local_peer_id = parachain_network.local_peer_id();
+		let public_addresses = public_addresses
+			.into_iter()
+			.filter_map(|mut addr| match addr.iter().last() {
+				Some(Protocol::P2p(peer_id)) if &peer_id == local_peer_id.as_ref() => {
+					addr.pop();
+					Some(addr)
+				},
+				Some(Protocol::P2p(_)) => {
+					warn!(
+						target: LOG_TARGET,
+						"Discarding public address containing not our peer ID: {addr}",
+					);
+					None
+				},
+				_ => Some(addr),
+			})
+			.collect();
+
+		Self {
+			para_id,
+			para_id_scale_compact: CompactRef(&para_id).encode(),
+			relay_chain_interface,
+			relay_chain_network,
+			current_epoch_key: None,
+			next_epoch_key: None,
+			current_epoch_publish_retry: Box::pin(Fuse::terminated()),
+			next_epoch_publish_retry: Box::pin(Fuse::terminated()),
+			request_receiver,
+			parachain_network,
+			advertise_non_global_ips,
+			parachain_genesis_hash,
+			parachain_fork_id,
+			public_addresses,
+		}
+	}
+
+	async fn current_epoch(&self, hash: Hash) -> RelayChainResult<Epoch> {
+		let res = self
+			.relay_chain_interface
+			.call_runtime_api("BabeApi_current_epoch", hash, &[])
+			.await?;
+		Decode::decode(&mut &*res).map_err(Into::into)
+	}
+
+	async fn next_epoch(&self, hash: Hash) -> RelayChainResult<Epoch> {
+		let res = self
+			.relay_chain_interface
+			.call_runtime_api("BabeApi_next_epoch", hash, &[])
+			.await?;
+		Decode::decode(&mut &*res).map_err(Into::into)
+	}
+
+	fn epoch_key(&self, randomness: Randomness) -> KademliaKey {
+		self.para_id_scale_compact
+			.clone()
+			.into_iter()
+			.chain(randomness.into_iter())
+			.collect::<Vec<_>>()
+			.into()
+	}
```

Comment:
nit: Maybe we can move these fns (and similar) to a common module to avoid code duplication?

Reply:
I also though about it, but that would be at the cost of not being able to refer to `self` and would result in some more boilerplate code...